### PR TITLE
[WIP] Use Typesafe config and reusable modules in the notifier app

### PR DESCRIPTION
**Pull request by alexwlchan**
A first step towards #2967. Based on #2745.

@kenoir I’m getting a NullPointerException from Guice.createInjector when I try to run this code – any ideas?

### CODE REVIEW

File: archive/notifier/src/main/scala/uk/ac/wellcome/platform/archive/notifier/Main.scala [hunk 1]
```diff
@@ -1,43 +1,74 @@
...
-  val injector: Injector = Guice.createInjector(
-    new AppConfigModule(args),
-    ConfigModule,
+object Main extends App {
+  val app = new NotifierApp()
+  app.run()
+}
+
+class NotifierApp() extends WellcomeApp {
+  val configModule = TypesafeConfigModule
```
Comment (alicefuzier) [inline archive/notifier/src/main/scala/uk/ac/wellcome/platform/archive/notifier/Main.scala]: maybe if you use `lazy val` here and in line 28? I think in general we should find solutions where we don't set uninitialised vals in traits to override them in extending classes. It's brittle and often leads to weird initialisation errors

File: archive/notifier/src/main/scala/uk/ac/wellcome/platform/archive/notifier/Main.scala [hunk 2]
```diff
@@ -1,43 +1,74 @@
...
+  }
+}
 
-    val result = app.run()
+trait WellcomeApp extends Logging {
+  val configModule: Configurable
+  val modules: List[AbstractModule]
 
-    Await.result(result, Duration.Inf)
+  def injector: Injector =
```
[the remark this answers is not in the document]
Comment (alexwlchan) [reply archive/notifier/src/main/scala/uk/ac/wellcome/platform/archive/notifier/Main.scala]: I moved them up, yep.

**Comment by alexwlchan** [comment]
This is the current stack trace. In particular, I’m managing to construct the injector, but whenever I do `injector.getInstance(classOf[Foo])`, I get a different null pointer exception:

```
*** Running 'docker run --tty --rm --volume /Users/alexwlchan/.aws:/root/.aws --env USER_DIR=/Users/alexwlchan -p 9001:9001 notifier:latest'
15:39:25.264 [main] INFO  u.a.w.p.archive.notifier.NotifierApp - Starting service.
15:39:25.270 [main] ERROR u.a.w.p.archive.notifier.NotifierApp - Fatal error:
java.lang.NullPointerException: null
	at uk.ac.wellcome.platform.archive.notifier.NotifierApp.run(Main.scala:41)
	at uk.ac.wellcome.platform.archive.notifier.WellcomeApp.$init$(Main.scala:67)
	at uk.ac.wellcome.platform.archive.notifier.NotifierApp.<init>(Main.scala:25)
	at uk.ac.wellcome.platform.archive.notifier.Main$.delayedEndpoint$uk$ac$wellcome$platform$archive$notifier$Main$1(Main.scala:21)
	at uk.ac.wellcome.platform.archive.notifier.Main$delayedInit$body.apply(Main.scala:20)
	at scala.Function0.apply$mcV$sp(Function0.scala:34)
	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App.$anonfun$main$1$adapted(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.App.main(App.scala:76)
	at scala.App.main$(App.scala:74)
	at uk.ac.wellcome.platform.archive.notifier.Main$.main(Main.scala:20)
	at uk.ac.wellcome.platform.archive.notifier.Main.main(Main.scala)
```

**Comment by alicefuzier** [comment]
Die Guice, Die!

**Comment by alexwlchan** [comment]
Closing in favour of #2974.

**Pull request closed without merging**